service/src/main/java/de/adorsys/psd2/sandbox/xs2a/service/redirect/RedirectService.java

```diff
@@ -0,0 +1,113 @@
+
+@Service
+@Transactional
+public class RedirectService {
+
+  private PisPaymentDataRepository pisPaymentDataRepository;
+  private PisAuthorizationRepository pisAuthorizationRepository;
+  private AisConsentAuthorizationRepository aisConsentAuthorizationRepository;
+  private AisConsentRepository aisConsentRepository;
+
+  /**
+   * Create a new RedirectService instance.
+   *
+   * @param pisPaymentDataRepository          PisPaymentRepository
+   * @param pisAuthorizationRepository        PisAuthorizationRepository
+   * @param aisConsentAuthorizationRepository AisConsentAuthorizationRepository
+   * @param aisConsentRepository              AisConsentRepository
+   */
+  public RedirectService(PisPaymentDataRepository pisPaymentDataRepository,
+      PisAuthorizationRepository pisAuthorizationRepository,
+      AisConsentAuthorizationRepository aisConsentAuthorizationRepository,
+      AisConsentRepository aisConsentRepository) {
+    this.pisPaymentDataRepository = pisPaymentDataRepository;
+    this.pisAuthorizationRepository = pisAuthorizationRepository;
+    this.aisConsentAuthorizationRepository = aisConsentAuthorizationRepository;
+    this.aisConsentRepository = aisConsentRepository;
+  }
+
+  /**
+   * Sets consent status in CMS depending on PSU-ID.
+   *
+   * @param externalId Consent Id
+   * @param psuId      Psu Id
+   */
+  public void handleConsentCreationRedirectRequest(String externalId, String psuId) {
+    Optional<AisConsentAuthorization> aisConsentAuthorization = aisConsentAuthorizationRepository
+        .findByExternalId(externalId);
+
+    if (!aisConsentAuthorization.isPresent()) {
+      //TODO handle error case
+      return;
+    }
+    AisConsentAuthorization aisConsentAuth = aisConsentAuthorization.get();
+
+    AisConsent consent = aisConsentRepository.findOne(aisConsentAuth.getId());
+    // TODO: Set status of PSU from TestDataService
+    consent.setConsentStatus(ConsentStatus.VALID);
+    aisConsentAuth.setScaStatus(ScaStatus.PSUAUTHENTICATED);
+
+    aisConsentRepository.save(consent);
+    aisConsentAuthorizationRepository.save(aisConsentAuth);
+  }
+
+  /**
+   * Sets payment status in CMS depending on PSU-ID.
+   *
+   * @param externalId Payment Id
+   * @param psuId      Psu Id
+   */
+  public void handlePaymentRedirectRequest(String externalId, String psuId, boolean isInit) {
+    Optional<PisAuthorization> pisAuthorization = pisAuthorizationRepository
+        .findByExternalId(externalId);
+
+    if (!pisAuthorization.isPresent()) {
+      //TODO handle error case
+      return;
+    }
+
+    PisAuthorization payment = pisAuthorization.get();
+
+    Optional<List<PisPaymentData>> pisPaymentDataList = pisPaymentDataRepository
+        .findByPaymentId(payment.getPaymentData().getPaymentId());
+
+    if (!pisPaymentDataList.isPresent()) {
+      //TODO handle error case
+      return;
+    }
+    PisPaymentData pisPaymentData = pisPaymentDataList.get().get(0);
```

Comment:
Is the list always not empty?

Reply:
There is a check for a non existing DataList before this function. So the function should always be called with a filled LIst. 

Please check and resolve the conversation when its okay.

Reply:
We check if there IS a list (not null) but not if its not empty. I've extended the check.